### DISCUSSION TITLE
Use react router in client components.

### DISCUSSION THREAD

**Discussion by carloslato** [honojs/hono#2348] [Q&A] [answered]
Hi guys, I am testing the client components and I would like to know if it is possible to use react libraries like React Router. I tried but I got some errors in console:

`Uncaught TypeError: d.current is null`

My goal is to deploy an app in cloudflare pages and use /functions for the backend. My other alternative to solve my problem is to simply use React for the frontend, but I wanted to know if there was a way to do it entirely with hono.

**Comment by Luzefiru** [comment, marked as answer]
I've got a sample repository with Hono on Cloudflare Pages found [here](https://github.com/Luzefiru/hono-react-tailwind).

### How it Works
The index path `/` [renders the React client](https://github.com/Luzefiru/hono-react-tailwind/blob/59c1f2cc5438bd89144163753bfdb3be77e33faf/src/index.tsx#L6) it simply calls [a script to hydrate the React Router](https://github.com/Luzefiru/hono-react-tailwind/blob/59c1f2cc5438bd89144163753bfdb3be77e33faf/src/client.tsx#L6) `<RouteSwitch />` component. This successfully maps wildcard routes when you navigate to them via URL and you can simply start coding a SPA-like experience in the [`<RouteSwitch />`](https://github.com/Luzefiru/hono-react-tailwind/blob/main/src/components/RouteSwitch.tsx) file.

After setting up the React script serving & router switching, you can add any routes or middleware just like you would any Hono app to the `index.tsx` file.

### For Improvement
If you have any feedback for it, feel free to raise an issue or make a PR as I'd like to improve on it for future reference. I hope this helps!

**Reply by carloslato** [reply]
Look great, thanks. I will check it

**Reply by ntk221** [reply]
In my case, I've used `HashRouter` instead of `BrowserRouter`.  This prevents additional server requests when switching paths in the browser.

**Reply by Luzefiru** [reply]
You're right @ntk221! That would save a ton of requests. Thank you for pointing that out.
